library/common/network/socket_tag_socket_option_impl.cc

```diff
@@ -0,0 +1,49 @@
+#include "library/common/network/socket_tag_socket_option_impl.h"
+
+#include "envoy/config/core/v3/base.pb.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/common/scalar_to_byte_vector.h"
+
+#include "library/common/jni/android_jni_utility.h"
+
+namespace Envoy {
+namespace Network {
+
+SocketTagSocketOptionImpl::SocketTagSocketOptionImpl(uid_t uid, uint32_t traffic_stats_tag)
+    : optname_(0, 0, "socket_tag"), uid_(uid), traffic_stats_tag_(traffic_stats_tag) {}
+
+bool SocketTagSocketOptionImpl::setOption(
+    Socket& socket, envoy::config::core::v3::SocketOption::SocketState state) const {
+  if (state != envoy::config::core::v3::SocketOption::STATE_PREBIND) {
+    return true;
+  }
+
+  int fd = socket.ioHandle().fdDoNotUse();
```

Comment:
Can we comment here why this is OK for multiplexed protocols (where many streams may share the connection)
Also API comments on any limitations (not sure if it'll crash or fail if used on internal socket, or any other without fd)

Reply:
Added a comment about multiplexed protocols. Thanks!

Can you say more about "internal sockets" or sockets without FDs?

Reply:
so it's not uncommon Envoy users want to
1) decap connect
2) L7 process decapsulated HTTP (imagine a full HTTP/2 connection in here)
3) re-encap with connect
If either the inner or outer stream can be HTTP/2, you accomplish this via 3 passes through Envoy: one L7 pass which decapsulated, and forwards the inner payload to a loopback listener which does L7 processing, and forwards raw (maybe HTTP/2) bytes to another loopback listener which encapsulates.  The kernel context switches get cumbersome and keeping track of things like original IP get fraught, so three's an "internal listener" which lets you bypass the kernel, and basically do it all through userspace instead.  in this case there's no fd, and I'm not sure if the fd() call returns  -1 or ENVOY_BUG but tl;dr it's worth commenting that it only works if we can access a raw fd.

Reply:
Ah, I see. Done.